File: backtracking/1415. The k-th Lexicographical String of All Happy Strings of Length n.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
    void gen_strings(
        string curr,
        const char last,
        vector<string> &store,
        const int &n,
        const int &k
    )  {
        if (store.size() == k) {
            return;
        }
        if (curr.size() == n) {
            store.push_back(curr);
            return;
        }
        
        static char chars[3] = {'a', 'b', 'c'};
        for (const auto &ch : chars) {
            if (ch != last) {
                curr.push_back(ch);
                gen_strings(curr, ch, store, n, k);
                curr.pop_back();
            }
        }
    }
public:
    string getHappyString(int n, int k) {
        vector<string> store;    
        this->gen_strings("", '\0', store, n, k);
    
        if (k <= store.size()) {
            return store[k - 1];
        } else {
            return "";
        }
    }
};

int main() {
    Solution run = Solution();
    cout << run.getHappyString(3, 9);
}

File: backtracking/1415. The k-th Lexicographical String of All Happy Strings of Length n.py
class Solution:
    def getHappyString(self, n: int, k: int) -> str:
        store = [] 
        def gen_strings(curr, last):
            if len(store) == k:
                return
            if len(curr) == n:
                store.append(curr)
                return

            for ch in ['a', 'b', 'c']:
                if ch != last:
                    gen_strings(curr + ch, ch)

        gen_strings("", "")
            
        return store[k - 1] if k <= len(store) else ""

if __name__ == "__main__":
    print(Solution().getHappyString(3, 9))
